client/allocrunner/taskrunner/envoy_version_hook_test.go

```diff
@@ -422,15 +422,72 @@ func TestTaskRunner_EnvoyVersionHook_Prestart_error(t *testing.T) {
 		TaskDir: allocDir.NewTaskDir(alloc.Job.TaskGroups[0].Tasks[0].Name),
 		TaskEnv: taskEnvDefault,
 	}
-	require.NoError(t, request.TaskDir.Build(false, nil))
+	must.NoError(t, request.TaskDir.Build(false, nil))
 
 	// Prepare a response
 	var response ifs.TaskPrestartResponse
 
 	// Run the hook, error should be recoverable
 	err := h.Prestart(context.Background(), request, &response)
-	require.EqualError(t, err, "error retrieving supported Envoy versions from Consul: some consul error")
+	must.ErrorContains(t, err, "error retrieving supported Envoy versions from Consul: some consul error")
 
 	// Assert the hook is not Done
-	require.False(t, response.Done)
+	must.False(t, h.done)
+}
+
+func TestTaskRunner_EnvoyVersionHook_Prestart_restart(t *testing.T) {
+	ci.Parallel(t)
+
+	logger := testlog.HCLogger(t)
+
+	// Setup an Allocation
+	alloc := mock.ConnectAlloc()
+	alloc.Job.TaskGroups[0].Tasks[0] = mock.ConnectSidecarTask()
+	allocDir, cleanupDir := allocdir.TestAllocDir(t, logger, "EnvoyVersionHook", alloc.ID)
+	defer cleanupDir()
+
+	// Set up a mock for Consul API.
+	mockProxiesAPI := consul.MockSupportedProxiesAPI{
+		Value: map[string][]string{
+			"envoy": {"1.15.0", "1.14.4"},
+		},
+		Error: nil,
+	}
+
+	// Run envoy_version hook
+	h := newEnvoyVersionHook(newEnvoyVersionHookConfig(alloc, mockProxiesAPI, logger))
+
+	// Create a prestart request
+	request := &ifs.TaskPrestartRequest{
+		Task:    alloc.Job.TaskGroups[0].Tasks[0],
+		TaskDir: allocDir.NewTaskDir(alloc.Job.TaskGroups[0].Tasks[0].Name),
+		TaskEnv: taskEnvDefault,
+	}
+	must.NoError(t, request.TaskDir.Build(false, nil))
+
+	// Prepare a response
+	var response ifs.TaskPrestartResponse
+
+	// Run the hook and ensure its internal state is marked as done and the
+	// tasks image has been modified.
+	must.NoError(t, h.Prestart(context.Background(), request, &response))
+	must.True(t, h.done)
+	must.Eq(t, "envoyproxy/envoy:v1.15.0", request.Task.Config["image"])
+
+	// Overwrite the previously modified image. This is the same behaviour that
+	// occurs when the server sends a non-destructive allocation update.
+	request.Task.Config["image"] = "${meta.connect.sidecar_image}"
+
+	// Run the PreKilling function which gets called when an allocation or the
+	// task is being restarted.
+	must.NoError(t, h.PreKilling(nil, nil, nil))
```

Comment:
Similarly to my comment above, we might want this test to exercise a non-requested task kill too.

Reply:
I have removed the addition of PreKilling as we are not tracking done anymore and made a slight modification to the test which mimics multiple calls.